frontend/packages/schema-model/src/lib/SchemaModel.ts

```diff
@@ -586,6 +585,92 @@ export class SchemaModel {
     newModel.getNodeMap().forEach((node) => this.nodeMap.set(node.schemaPointer, node));
     return this;
   }
+
+  public willResultInCircularReferences(
+    childSchemaPointer: string,
+    parentSchemaPointer: string,
+  ): boolean {
+    const childDefinitions = this.listDefinitionsWithin(childSchemaPointer);
+    const parentDefinitions = this.listAllDefinitionsAbove(parentSchemaPointer);
+    return ArrayUtils.hasIntersection(childDefinitions, parentDefinitions);
+  }
+
+  private listDefinitionsWithin(schemaPointer: string): string[] {
+    const allNodesWithin = this.listAllNodesWithin(schemaPointer);
+    return this.filterDefinitions(allNodesWithin);
+  }
+
+  private listAllNodesWithin(schemaPointer: string): UiSchemaNode[] {
+    const node = this.getNodeBySchemaPointer(schemaPointer);
+    const allNodesWithin: UiSchemaNode[] = [node];
+    const children = this.getChildNodes(schemaPointer);
+    children.forEach((child) => {
+      allNodesWithin.push(...this.listAllNodesWithin(child.schemaPointer));
+    });
+    return allNodesWithin;
+  }
+
+  private filterDefinitions(nodes: UiSchemaNodes): string[] {
+    const directDefinitions = this.filterDirectDefinitions(nodes);
+    const refererredDefinitions = this.filterReferredDefinitions(nodes);
+    return [...directDefinitions, ...refererredDefinitions];
+  }
+
+  private filterDirectDefinitions(nodes: UiSchemaNodes): string[] {
+    return nodes.filter((node) => this.isDefinitionRoot(node)).map((node) => node.schemaPointer);
+  }
+
+  private isDefinitionRoot(node: UiSchemaNode): boolean {
+    return isDefinition(node) && this.isChildOfRoot(node);
+  }
+
+  private isChildOfRoot(node: UiSchemaNode): boolean {
+    const parent = this.getParentNode(node.schemaPointer);
+    return parent && parent.schemaPointer === ROOT_POINTER;
+  }
```

Comment:
Bare av nysgjerrighet, finnes ikke disse funksjonene på node-klassen fra før?

Reply:
Jeg syns datamodell-klassen er veldig komplisert altså, så veldig mulig at jeg bare ikke henger helt med 😅

Reply:
Det gjør de ikke. `isDefinition` er egentlig litt upresis i navnet, for den returnerer også `true` for alle elementer som er underelementer av typer. Det den sjekker er kun om pekeren starter med `#/$defs`. Vi burde gi den et bedre navn, men endringen ville blitt uforholdsmessig stor om jeg gjorde det nå.

Ja, denne klassen begynner å bli altfor stor. Vi bør absolutt gjøre en jobb for å dele den opp i klasser med mer spesifikke ansvarsområder. I hvert fall er den bedre enn spaghettikoden vi hadde før.

Reply:
De nye, private metodene jeg har lagt til brukes egentlig bare av `willResultInCircularReferences`. Jeg skal se om jeg finner en god måte å isolere dette på.

Reply:
Da har jeg opprettet [et nytt endringsforslag som legger et grunnlag for refaktoreringen](https://github.com/Altinn/altinn-studio/pull/13812).

Reply:
Nå har jeg oppdatert koden med en egen klasse for disse metodene.